include/wil/com_apartment_variable.h

```diff
@@ -0,0 +1,234 @@
+//*********************************************************
+//
+//    Copyright (c) Microsoft. All rights reserved.
+//    This code is licensed under the MIT License.
+//    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF
+//    ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
+//    TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+//    PARTICULAR PURPOSE AND NONINFRINGEMENT.
+//
+//*********************************************************
+#ifndef __WIL_COM_APARTMENT_VARIABLE_INCLUDED
+#define __WIL_COM_APARTMENT_VARIABLE_INCLUDED
+
+#include <unordered_map>
+#include <any>
+#include <type_traits>
+#include <functional>
+#include "cppwinrt.h"
+#include <roapi.h>
+#include <objidl.h>
+#include "result_macros.h"
+
+#ifndef WIL_ENABLE_EXCEPTIONS
+#error This header requires exceptions
+#endif
+
+namespace wil
+{
+    // Determine if apartment variables are supported in the current process context.
+    // Prior to build 22365, the APIs needed to create apartment variables (e.g. RoGetApartmentIdentifier)
+    // failed for unpackaged processes. For MS people, see http://task.ms/31861017 for details.
+    // APIs needed to implement apartment variables did not work in non-packaged processes.
+    inline bool are_apartment_variables_supported()
+    {
+        unsigned long long apartmentId{};
+        return RoGetApartmentIdentifier(&apartmentId) != HRESULT_FROM_WIN32(ERROR_API_UNAVAILABLE);
+    }
+
+    namespace details
+    {
+        struct apt_var_platform
+        {
+            static unsigned long long GetApartmentId()
+            {
+                unsigned long long apartmentId{};
+                FAIL_FAST_IF_FAILED(RoGetApartmentIdentifier(&apartmentId));
+                return apartmentId;
+            }
+
+            static void RegisterForApartmentShutdown(IApartmentShutdown* observer)
+            {
+                unsigned long long id{};
+                APARTMENT_SHUTDOWN_REGISTRATION_COOKIE cookie{}; // leaked, COM run down releases all instances
+                THROW_IF_FAILED(RoRegisterForApartmentShutdown(observer, &id, &cookie));
+            }
+        };
+
+        template<typename test_hook = apt_var_platform>

```

Comment:
Maybe call this `apartment_source` instead?

Reply:
its not a source of apartments

Reply:
other ideas? I'm going to won't fix.

Reply:
This is dependency injection, so `dependency` would be a not unreasonable name.